go/bundle/signature/signer.go

```diff
@@ -0,0 +1,193 @@
+			keyE.EncodeTextString("expires")
+			valueE.EncodeInt(s.Expires.Unix())
+		}),
+		cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+			keyE.EncodeTextString("subset-hashes")
+			subsetHashes := []*cbor.MapEntryEncoder{}
+			for url, rh := range s.SubsetHashes {
+				subsetHashes = append(subsetHashes,
+					cbor.GenerateMapEntry(func(keyE *cbor.Encoder, valueE *cbor.Encoder) {
+						keyE.EncodeTextString(url)
+						valueE.EncodeArrayHeader(1 + len(rh.Hashes)*2)
+						valueE.EncodeTextString(rh.VariantsValue)
+						for _, ri := range rh.Hashes {
+							valueE.EncodeByteString(ri.HeaderSha256)
+							valueE.EncodeTextString(ri.PayloadIntegrityHeader)
+						}
+					}))
+			}
+			valueE.EncodeMap(subsetHashes)
+		}),
+	}
+	var buf bytes.Buffer
+	enc := cbor.NewEncoder(&buf)
+	if err := enc.EncodeMap(mes); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func NewSigner(ver version.Version, certs certurl.CertChain, privKey crypto.PrivateKey, validityUrl *url.URL, date time.Time, duration time.Duration) (*Signer, error) {
+	if ver == version.Unversioned {
+		return nil, errors.New("signature: unversioned bundles cannot be signed")
+	}
+
+	if err := certs.Validate(); err != nil {
+		return nil, err
+	}
+	authSha256 := certs[0].CertSha256()
+
+	return &Signer{
+		Version: ver,
+		Certs:   certs,
+		PrivKey: privKey,
+		Rand:    rand.Reader,
+		SignedSubset: SignedSubset{
+			ValidityUrl:  validityUrl,
+			AuthSha256:   authSha256,
+			Date:         date,
+			Expires:      date.Add(duration),
+			SubsetHashes: make(map[string]*ResponseHashes),
+		},
+	}, nil
+}
+
+// CanSignForURL returns true iff this signer can sign for a resource with given URL.
+func (s *Signer) CanSignForURL(u *url.URL) bool {
+	return s.Certs[0].Cert.VerifyHostname(u.Hostname()) == nil
+}
+
+// AddExchange adds resource integrity information of e for signing.
+func (s *Signer) AddExchange(e *bundle.Exchange, payloadIntegrityHeader string) error {
+	headerSha256, err := e.Response.HeaderSha256()
+	if err != nil {
+		return err
+	}
+	ri := &ResourceIntegrity{HeaderSha256: headerSha256, PayloadIntegrityHeader: payloadIntegrityHeader}
+
+	if _, ok := s.SubsetHashes[e.Request.URL.String()]; ok {
+		// TODO: Fix this when we implement variants.
+		return errors.New("signature: multiple exchanges for single URL is not supported")
+	}
+
+	s.SubsetHashes[e.Request.URL.String()] = &ResponseHashes{
+		VariantsValue: "",
+		Hashes:        []*ResourceIntegrity{ri},
+	}
+	return nil
+}
+
+// UpdateSignatures updates bundle.Signatures by adding the cert chain and
+// the signature of exchanges added with AddExchange.
+func (s *Signer) UpdateSignatures(signatures *bundle.Signatures) (*bundle.Signatures, error) {
+	if signatures == nil {
+		signatures = &bundle.Signatures{}
+	}
+
+	authorityIndex := len(signatures.Authorities)
+	signatures.Authorities = append(signatures.Authorities, s.Certs...)
```

Comment:
I think this is ok, but it may lead to duplicate intermediate certificates in the authorities list.

Reply:
Added a TODO comment to deduplicate intermediate certificates.